res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,3066 @@
+                this.rightVelocityFactor = averageFactor;
+            }
+        }
+
+        if (velocityLeft) {
+            velocityLeft += wheelAbsoluteMax / 2;
+        }
+
+        if (velocityRight) {
+            velocityRight += wheelAbsoluteMax / 2;
+        }
+
+        velocityLeft = Math.min(
+            MaxWheelForce,
+            Math.floor(velocityLeft)
+        );
+
+        velocityRight = Math.min(
+            MaxWheelForce,
+            Math.floor(velocityRight)
+        );
+
+        motorData[3] = velocityLeft & 0xff;
+        motorData[4] = velocityLeft >> 8;
+
+        motorData[8] = velocityRight & 0xff;
+        motorData[9] = velocityRight >> 8;
+        controller.send(motorData, null, 49, true);
+    }
+    incomingData(data, _, forceReportId) {
+        const reportId = forceReportId || data[0];
+        if (reportId === 1) {
+            this.inPackets[1].handleInput(data.buffer, !!forceReportId);
+        } else if (reportId === 2) {
+            this.inPackets[2].handleInput(data.buffer, !!forceReportId);
+            // The master volume, booth volume, headphone mix, and headphone volume knobs
+            // control the controller's audio interface in hardware, so they are not mapped.
+        } else if (reportId === 3) {
+            // The 32 bit unsigned ints at bytes 8 and 36 always have exactly the same value,
+            // so only process one of them. This must be processed before the wheel positions.
+            const oldWheelTimer = wheelTimer;
+            const view = new DataView(data.buffer);
+            wheelTimer = view.getUint32(8, true);
+            // Processing first value; no previous value to compare with.
+            if (oldWheelTimer === null) {
+                return;
+            }
+            wheelTimerDelta = wheelTimer - oldWheelTimer;
+            if (wheelTimerDelta < 0) {
+                wheelTimerDelta += wheelTimerMax;
+            }
+
+            this.leftDeck.wheelRelative.input(view.getUint16(12, true));
+            this.rightDeck.wheelRelative.input(view.getUint16(40, true));
+        } else {
+            console.warn("Unsupported HID repord with ID "+ reportId + ". Contains: "+data);
+        }
+    }
+    init() {
+        // sending these magic packets is required for the jog wheel LEDs to work
+        const wheelLEDinitPacket = Array(26).fill(0);
+        wheelLEDinitPacket[1] = 1;
+        wheelLEDinitPacket[2] = 3;
+        controller.send(wheelLEDinitPacket, null, 48, true);
+        wheelLEDinitPacket[0] = 1;
+        controller.send(wheelLEDinitPacket, null, 48);
+
+        // Init wheel timer data
+        wheelTimer = null;
+        wheelTimerDelta = 0;
+
+        // get state of knobs and faders
+        this.incomingData(new Uint8Array(controller.getInputReport(1)), null, 1);
+        this.incomingData(new Uint8Array(controller.getInputReport(2)), null, 2);
```

Comment:
I was unable to make the spread operator work, this is why I have introduced this `offset` later. I will give it a go again, and report back whether or not it works now.

Reply:
Yes, I can confirm that it still doesn't work, although this is is valid syntax and it works in a node console (with a dummy array for `controller.getInputReport`).

It creates a `Uint8Array` of size 1 with the first byte being the reportID, so either the spread operator isn't supported in the JS engine, or `controller.getInputReport` doesn't return a real  `Uint8Array`.

Not sure how to resolve that, but I am pretty sure this isn't a S4-specific bug. I will keep the nasty `offset` hack for now although I agree with you, it looks ugly and would be great to be able to make it work!

Reply:
The following code works (but is unecessary complicated, because it contains a workaround for an issue which doesn't exist here):
https://github.com/mixxxdj/mixxx/blob/dc4cc198b238cb95fac8c6f80d888c24e8856a98/res/controllers/Traktor-Kontrol-Z2-hid-scripts.js#L775-L784

Reply:
Right, so wrapping it into an other `new Uint8Array` seems to do the trick. 

```javascript
        this.incomingData(new Uint8Array([0x01, ...Uint8Array.from(new Uint8Array(controller.getInputReport(0x01)))]));
        this.incomingData(new Uint8Array([0x02, ...Uint8Array.from(new Uint8Array(controller.getInputReport(0x02)))]));
```

That doesn't look right tho, and still feels like a bug to me :) 
But that's surely better than the offset solution.

Reply:
> That doesn't look right tho, and still feels like a bug to me :)
> But that's surely better than the offset solution.

I think approach to call this.incomingData results in this ugly code. To make it clean, handleInput should get the ReportId seperated as extra argument. Than you can call it with the unmodified array from controller.getInputReport.

Consider to change
https://github.com/mixxxdj/mixxx/blob/4e1115fc9795d644fba329cf0a30b1351aeb8043/res/controllers/Traktor-Kontrol-S4-MK3.js#L3039-L3044
to something like
```
 incomingData(data) { 
     const reportId = data[0]; 
     if (reportId === 1) { 
         this.inReports[1].handleInput(reportId, data.buffer.slice(1)); 
     } else if (reportId === 2) { 
         this.inReports[2].handleInput(reportId, data.buffer.slice(1)); 
```